src/security/acl_api.c

```diff
@@ -0,0 +1,450 @@
+
+	D_ALLOC(*new_acl, sizeof(struct daos_acl) + new_len);
+	if (*new_acl == NULL) {
+		return -DER_NOMEM;
+	}
+
+	(*new_acl)->dal_ver = acl->dal_ver;
+	(*new_acl)->dal_len = new_len;
+
+	copy_acl_with_new_ace_inserted(acl, *new_acl, new_ace);
+
+	return DER_SUCCESS;
+}
+
+static bool
+type_is_valid(enum daos_acl_principal_type type)
+{
+	bool result = false;
+
+	switch (type) {
+	case DAOS_ACL_USER:
+	case DAOS_ACL_GROUP:
+	case DAOS_ACL_OWNER:
+	case DAOS_ACL_OWNER_GROUP:
+	case DAOS_ACL_EVERYONE:
+		result = true;
+		break;
+	}
+
+	return result;
+}
+
+static bool
+type_needs_name(enum daos_acl_principal_type type)
+{
+	/*
+	 * The only ACE types that require a name are User and Group. All others
+	 * are "special" ACEs that apply to an abstract category.
+	 */
+	if (type == DAOS_ACL_USER || type == DAOS_ACL_GROUP) {
+		return true;
+	}
+
+	return false;
+}
+
+static bool
+principal_meets_type_requirements(enum daos_acl_principal_type type,
+		const char *principal_name, size_t principal_name_len)
+{
+	return	(!type_needs_name(type) ||
+		(principal_name != NULL && principal_name_len != 0));
+}
+
+int
+daos_acl_remove_ace_realloc(struct daos_acl *acl,
+		enum daos_acl_principal_type type, const char *principal_name,
+		size_t principal_name_len, struct daos_acl **new_acl)
+{
+	struct daos_ace	*current;
+	uint8_t		*pen;
+
+	if (acl == NULL || new_acl == NULL || !type_is_valid(type) ||
+		!principal_meets_type_requirements(type, principal_name,
+				principal_name_len)) {
+		return -DER_INVAL;
+	}
+
+	if (daos_acl_get_ace_for_principal(acl, type, principal_name) == NULL) {
+		/* requested principal not in the list */
+		return -DER_NONEXIST;
+	}
+
+	D_ALLOC(*new_acl, sizeof(struct daos_acl) + acl->dal_len);
+	if (*new_acl == NULL) {
+		return -DER_NOMEM;
+	}
+
+	(*new_acl)->dal_ver = acl->dal_ver;
+
+	pen = (*new_acl)->dal_ace;
+	current = daos_acl_get_first_ace(acl);
+	while (current != NULL) {
+		if (!ace_matches_principal(current, type, principal_name,
+				principal_name_len)) {
+			pen = write_ace(current, pen);
+			(*new_acl)->dal_len += daos_ace_get_size(current);
+		}
+
+		current = daos_acl_get_next_ace(acl, current);
+	}
+
+	return DER_SUCCESS;
+}
+
+struct daos_ace *
+daos_acl_get_first_ace(struct daos_acl *acl)
+{
+	if (acl == NULL || acl->dal_len == 0) {
+		return NULL;
+	}
+
+	return (struct daos_ace *)acl->dal_ace;
+}
+
+static bool
+is_in_ace_list(uint8_t *addr, struct daos_acl *acl)
+{
+	uint8_t *start_addr = acl->dal_ace;
+	uint8_t *end_addr = acl->dal_ace + acl->dal_len;
+
+	return addr >= start_addr && addr < end_addr;
+}
+
+struct daos_ace *
+daos_acl_get_next_ace(struct daos_acl *acl, struct daos_ace *current_ace)
+{
+	size_t offset;
+
+	if (acl == NULL || current_ace == NULL) {
+		return NULL;
+	}
+
+	/* already at/beyond the end */
+	if (!is_in_ace_list((uint8_t *)current_ace, acl)) {
+		return NULL;
+	}
+
+	/* there is no next item */
+	offset = sizeof (struct daos_ace) + current_ace->dae_principal_len;
```

Comment:
(style)  space prohibited between function name and open parenthesis '('

Reply:
Fixed